dali/python/nvidia/dali/plugin/pytorch/__init__.py

```diff
@@ -224,7 +224,10 @@ def __init__(
                 # here we should set if to False again
                 self._ever_consumed = False
             except StopIteration:
-                assert False, (
+                # This case might not be an error if we're iterating over pipeline that is
+                # currently at the end of epoch, for example because it was restored from
+                # checkpoint.
+                assert any(not p._first_iter for p in self._pipes), (
                     "It seems that there is no data in the pipeline. This may happen "
                     "if `last_batch_policy` is set to PARTIAL and the requested batch size is "
                     "greater than the shard size."
```

Comment:
Pytorch iterator has a check for non-empty pipeline. This may fail when a checkpoint was saved after the last iteration. Then, after restoring, we get a pipeline that will throw `StopIteration` at the very beginning, but isn't a reason to worry. Are we ok with this fix?

Reply:
As for the check itself: should we make it conditional based on some variable that indicates that we are not running from scratch but from a checkpoint? 

We would need to consume the first `StopIteration` so we still have something in `_first_batch`, otherwise, won't this be broken? 

Or we just need to remember that fact, and reraise `StopIteration` from the `__next__`, so the user calls .Reset() and starts from the next batch? <- I think this may work, but is a bit on the spaghetti side. Still, we probably don't want to raise StopIteration from `__init__`.

Reply:
> We would need to consume the first StopIteration so we still have something in _first_batch, otherwise, won't this be broken?

We don't need to fill `_first_batch`. Iterator's `__next__` is opportunistic in that aspect: if there's something in `_first_batch` it's returned, but otherwise the iterator just proceeds the usual way: https://github.com/NVIDIA/DALI/blob/e496263266dce5aa1dc99d4a17208864f277968f/dali/python/nvidia/dali/plugin/pytorch/__init__.py#L238-L241

Reply:
> As for the check itself: should we make it conditional based on some variable that indicates that we are not running from scratch but from a checkpoint?

Good idea, I added `is_restored_from_checkpoint` to `Pipeline` in https://github.com/NVIDIA/DALI/pull/5213/commits/f2c648ca31b7414efc8ca95eb71c82109c0c84d6